packages/hyperion-autologging/src/ALHeartbeat.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates. All Rights Reserved.
+ */
+
+'use strict';
+
+import type { Channel } from "@hyperion/hook/src/Channel";
+import { TimedTrigger } from '@hyperion/hyperion-util/src/TimedTrigger';
+import performanceAbsoluteNow from '@hyperion/hyperion-util/src/performanceAbsoluteNow';
+import * as Types from "@hyperion/hyperion-util/src/Types";
+import { ALChannelUIEvent } from "./ALUIEventPublisher";
+import * as ALEventIndex from "./ALEventIndex";
+import { ALLoggableEvent } from "./ALType";
+
+export enum ALHeartbeatType {
+  REGAIN_PAGE_VISIBILITY,
+  SCHEDULED,
+  START,
+}
+
+export type AdsALHeartbeatEventData = Readonly<
+  ALLoggableEvent &
+  {
+    event: "heartbeat";
+    heartbeatType: ALHeartbeatType;
+  }>;
+
+export type ALChannelHeartbeatEvent = Readonly<{
+  al_heartbeat: [AdsALHeartbeatEventData],
+}>;
+
+export type ALHeartbeatChannel = Channel<ALChannelHeartbeatEvent & ALChannelUIEvent>;
+
+export type InitOptions = Types.Options<{
+  channel: ALHeartbeatChannel;
+  heartbeatInterval?: number;
+  maxUserInactivityDuration?: number;
+}>;
+
+const HEARTBEAT_INTERVAL = 30 * 1000 /* DateConsts.MS_PER_SEC */;
+const MAX_USER_INACTIVITY_DURATION = 4 * HEARTBEAT_INTERVAL;
+const VISIBILITY_CHANE_EVENT = "visibilitychange";
+
+let _lastHeartbeatTime: number = 0;
+let _lastUserActionTime: number = performanceAbsoluteNow();
+let _timedLogger: TimedTrigger | null = null;
+let _options: InitOptions | null = null;
+let _releaseListeners: (() => void) | null;
+
+
+export function getInterval(): number {
+  return _options?.heartbeatInterval ?? HEARTBEAT_INTERVAL;
+}
+
+export function getLastHeartbeatTime(): number {
+  return _lastHeartbeatTime;
+}
+
+export function getLastUserActionTime(): number {
+  return _lastUserActionTime;
+}
+
+function isActive(): boolean {
+  return _timedLogger != null;
+}
+
+export function start(options: InitOptions): void {
+  if (isActive()) {
+    return;
+  }
+  _options = options;
+  const { channel } = options;
+
+  const userActionListener = channel.addListener(
+    'al_ui_event',
+    ({ eventTimestamp }) => {
+      _lastUserActionTime = eventTimestamp;
+    },
+  );
+
+  const pageVisibilityListener = (_e: Event) => {
+    const isHidden = document.hidden;
```

Comment:
WWW also checks for `(moz|ms|webkit)Hidden` (referencing `html/shared/core/Visibility.js`) - should we be doing the same here?

Reply:
Yes, I checked it but that seems to be pretty old. All modern browsers do support this already .

https://caniuse.com/?search=visibilitychange
https://caniuse.com/mdn-api_document_hidden

That's why I decided to with a simpler code.